### DISCUSSION TITLE
Using a std::map instead of language.json file

### DISCUSSION THREAD

**Discussion by R-Abbasi** [skunkforce/OmniView#97] [Ideas]
This is a first step to using a `std::map` rather than `language.json` file utilized in my current `pr`. `std::map`s functionalities seem easier to treat, more beneficial, and reducing codebase in general:
```
//  ...
  static std::map<const char*, const char*> englishLan {
        {"Menu", "Menu"},
        {"LanOption", "Language Option"},
        {"English", "English"},
        {"Settings", "Settings"},
        {"Reset", "Reset"},
        {"Version", "Version"},
        {"Diagnostics", "Diagnostics"},
        {"Compression", "Compression"},
        {"Anlyz_crnt_waveform", "Analyze current waveform"},
        {"Gnrt_trning_data", "Generate training data"},
        {"Timing-Belt", "Timing-Belt"},
        {"Fuel-Delivery-Pump", "Fuel-Delivery-Pump"},
        {"Common-Rail-Pressure", "Common-Rail-Pressure"},
        {"Help", "Help"},
        {"HelpLink", "Click on the button to go to the help page"},
        {"Dvc_search", "Search for\nDevices"},
        {"Save", "Save"},
        {"Save warning", "Save warning"},
        {"No_dvc_available", "No devices are available ..."},
        {"AnalyzeData", "Analyze\nData"},
        {"Crt_trng_data", "Create\nTraining\nData"},
        {"Back", "Back"} };

   static std::map<const char*, const char*> germanLan {
       {"Menu", "Menü"},
       {"LanOption", "Sprachoption"},
       {"English", "Englisch"},
       {"Settings", "Einstellungen"},
       {"Reset", "Zurücksetzen"},
       {"Version", "Ausführung"},
       {"Diagnostics", "Diagnose"},
       {"Compression", "Kompression"},
       {"Anlyz_crnt_waveform", "Aktuelle Wellenform analysieren"},
       {"Gnrt_trning_data", "Trainingsdaten generieren"},
       {"Timing-Belt", "Zahnriemen"},
       {"Fuel-Delivery-Pump", "Kraftstoff-Förderpumpe"},
       {"Common-Rail-Pressure", "Common-Rail-Druck"},
       {"Help", "Hilfe"},
       {"HelpLink", "Klicken Sie auf den Button um zur Hilfeseite zu gelangen"},
       {"Dvc_search", "Suche\ngeräte"},
       {"Save", "speichern"},
       {"Save warning", "Speicherwarnung"},
       {"No_dvc_available", "Es sind keine Geräte verfügbar ..."},
       {"AnalyzeData", "Daten\nanalysieren"},
       {"Crt_trng_data", "Erstellen\nSie\nTrainingsdaten"},
       {"Back", "Zurück"} };
// ...
```
To see the consequence on decreasing code line on the one hand and increasing readability on the other hand, take a look at these two lines:

`if (ImGui::BeginMenu(appLanguage["Help"]))`
 
```
if (ImGui::BeginMenu(
            load_json<std::string>(language, "menubar", "help", "label")
                .c_str()))
```
Using a `std::map` initialized at compile-time compared to loading a `.json` file several times at run-time could lead to much better performance not only on small devices with a Linux distro - capable of running the project - but also on PCes, I assume.

**Comment by MaxClerkwell** [comment]
Maybe I am mistaken here, but this implies, that no changes in the language-file - or adding a language-file - is possible after it has been compiled. 

Can't we find a common ground here?
I really believe, that a language file shouldn't be compiled into the software, but has to be loaded at startup. 

Maybe loading the `language.json` at startup into a map could be sort of the situation or am I mistaken here?

**Comment by R-Abbasi** [comment]
The maps have some memory allocation at run-time but I don't see if they have something to do with language files, e.g., `Deutsch.json`. 
The idea of using `std::map`s implies omitting language files totally plus part of `config.json`. We can set maps for a common set of languages and add more in the next versions of the program when needed.

With a little modification, as the next step, we can even make it better:

```
enum class Key {
   Menu, 
   LanOption,
   ...
};
static const std::map<Key, const char*> englishLan {
        {Key::Menu, "Menu"},
        {Key::LanOption, "Language Option"},
        ...
};
```

With `.json` files instead, we need to load them from disk and then access their values. Still accessing the maps values sound to be way faster - they're sorted by default.
